onnxruntime/core/optimizer/pad_fusion.cc

```diff
@@ -115,18 +146,44 @@ Status PadFusion::Apply(Graph& graph, Node& pad_node, RewriteRuleEffect& rule_ef
   }
 
   Node& child_node = *graph.GetNode(pad_node.OutputNodesBegin()->Index());
-  auto child_pads = child_node.GetMutableAttributes()["pads"].mutable_ints();
-  uint32_t child_pads_size = static_cast<uint32_t>(child_pads->size());
 
-  for (uint32_t pads_index = 2, child_index = 0; pads_index < pads_size / 2; pads_index++, child_index++) {
-    child_pads->Set(child_index, child_pads->Get(child_index) + pads_values[pads_index]);
-    uint32_t mirrored_child_index = child_index + (child_pads_size / 2);
-    uint32_t mirrored_pad_index = pads_index + (pads_size / 2);
-    child_pads->Set(mirrored_child_index, child_pads->Get(mirrored_child_index) + pads_values[mirrored_pad_index]);
+  if (child_node.OpType() == "Cast") {
+    if (pad_node.SinceVersion() >= 11) {
+      if (pad_node.InputDefs().size() > 2) {
+        auto* pad_constant_value_proto = graph_utils::GetConstantInitializer(graph, pad_node.InputDefs()[2]->Name());
+        ONNX_NAMESPACE::TensorProto_DataType cast_data_type = static_cast<ONNX_NAMESPACE::TensorProto_DataType>(child_node.GetAttributes().at("to").i());
+        Initializer new_pad_constant_value{cast_data_type, pad_constant_value_proto->name(), pad_constant_value_proto->dims()};
+        // Create new initializers of Pad
+        ONNX_NAMESPACE::TensorProto new_pad_constant_value_tensor_proto;
+        new_pad_constant_value.ToProto(new_pad_constant_value_tensor_proto);
+
+        // Replace initializers of Pad node
+        graph.RemoveInitializedTensor(pad_node.InputDefs()[2]->Name());
```

Comment:
This can only be done if the initializer is unique to Pad right? what if that initializer is shared with some other node. Is a check needed for that?

Reply:
Removed the casting of Pad initializer because that is not required at all.